module/builder/collection/builder.go

```diff
@@ -60,277 +53,244 @@ func NewBuilder(
+	}
 
-		// TODO (ramtin): enable this again
-		// b.tracer.FinishSpan(parentID, trace.COLBuildOnSetup)
-		// b.tracer.StartSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
-		// defer b.tracer.FinishSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
-
-		// RATE LIMITING: the builder module can be configured to limit the
-		// rate at which transactions with a common payer are included in
-		// blocks. Depending on the configured limit, we either allow 1
-		// transaction every N sequential collections, or we allow K transactions
-		// per collection.
-
-		// keep track of transactions in the ancestry to avoid duplicates
-		lookup := newTransactionLookup()
-		// keep track of transactions to enforce rate limiting
-		limiter := newRateLimiter(b.config, parent.Height+1)
-
-		// look up previously included transactions in UN-FINALIZED ancestors
-		ancestorID := parentID
-		clusterFinalID := clusterFinal.ID()
-		for ancestorID != clusterFinalID {
-			ancestor, err := b.clusterHeaders.ByBlockID(ancestorID)
-			if err != nil {
-				return fmt.Errorf("could not get ancestor header (%x): %w", ancestorID, err)
-			}
-
-			if ancestor.Height <= clusterFinal.Height {
-				return fmt.Errorf("should always build on last finalized block")
-			}
-
-			payload, err := b.payloads.ByBlockID(ancestorID)
-			if err != nil {
-				return fmt.Errorf("could not get ancestor payload (%x): %w", ancestorID, err)
-			}
-
-			collection := payload.Collection
-			for _, tx := range collection.Transactions {
-				lookup.addUnfinalizedAncestor(tx.ID())
-				limiter.addAncestor(ancestor.Height, tx)
-			}
-			ancestorID = ancestor.ParentID
-		}
+	// TODO (ramtin): enable this again
+	// b.tracer.FinishSpan(parentID, trace.COLBuildOnSetup)
+	// b.tracer.StartSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
+	// defer b.tracer.FinishSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
+
+	// STEP TWO: create a lookup of all previously used transactions on the
+	// part of the chain we care about. We do this separately for
+	// un-finalized and finalized sections of the chain to decide whether to
+	// remove conflicting transactions from the mempool.
+
+	// keep track of transactions in the ancestry to avoid duplicates
+	lookup := newTransactionLookup()
+	// keep track of transactions to enforce rate limiting
+	limiter := newRateLimiter(b.config, parent.Height+1)
+
+	// RATE LIMITING: the builder module can be configured to limit the
+	// rate at which transactions with a common payer are included in
+	// blocks. Depending on the configured limit, we either allow 1
+	// transaction every N sequential collections, or we allow K transactions
+	// per collection.
+
+	// first, look up previously included transactions in UN-FINALIZED ancestors
+	fmt.Println("cluster chain final height: ", clusterChainFinalizedBlock.Height, clusterChainFinalizedBlock.ID())
+	err = b.populateUnfinalizedAncestryLookup(parentID, clusterChainFinalizedBlock.Height, lookup, limiter)
```

Comment:
why do we need to distinguish `populateUnfinalizedAncestryLookup` and `populateFinalizedAncestryLookup `?

Are they guaranteed to be connected? More importantly, would there be a gap between the last block of `populateFinalizedAncestryLookup` and the start block of `populateUnfinalizedAncestryLookup`? In this case, some transations will not be missed for dedup. 

I wonder why not computing the the `minRefFinalizedHeight` based on `minPossibleRefHeight` and the new index. So that we can just iterate from parentID all the way to the `minRefFinalizedHeight`, which is guaranteed to cover the range of blocks without any gap.